### DISCUSSION TITLE
Pendrive's filesystem destroyed after disk slot change

### DISCUSSION THREAD

**Discussion by slajerek** [keirf/flashfloppy#610] [Q&A]
Not filling a bug report yet and trying to figure out first if that is not a problem on my side. In short, I very often have a destroyed filesystem after quickly changing selected disk slot (meaning it is not readable by Gotek and PC anymore). 

I have Gotek SFRC922D connected to Amiga with 2 buttons and simple LCD. Previously had firmware v0.9.16a and thought that this may be an error thus upgraded to v3.29. I was using HxC file selector, now FF's

How to reproduce?
1. Select slot 000, boot Amiga.
2. Setup images on slot 001, 002, 003, etc. 
3. Select high-number slot using buttons, for example 010
4. Reboot Amiga, run game etc. 
5. When game is running (possibly even saving data...) quickly using buttons change slot from 010 iteratively back to 000, ie: 009, 008, 007...
6. Quite often Gotek during changing slot will blink the LCD and then go to F-F and can't boot anymore.
7. Put pendrive to PC - no filesystem existing, MBR broken (?) etc. Needs formatting.

Also this happened to me when I booted Amiga and was on slot 006, then directly after booting quickly changed to 000 and... boom! 

In overall I understand that changing selected disk slot when Amiga writes data may destroy disk image, but did not expect that can destroy whole filesystem. Also I have a question, how to properly change slots (i.e. to select slot 000 quickly?)

**Comment by keirf** [comment]
This shouldnt destroy the MBR. Have you tried other usb drives?

**Comment by manga303** [comment]
i got the same on diffrent usb drives, all 32fat formatet... some times it works fine on my amiga but then after i change some disks over 000. it make often A-D then over night, next day = drive error F-F. Stick on PC always the same partition is complete gone... re-partition, format, copy again the files over.... it is not only annyoing, all my save games etc. are gone again.... it makes no fun anymore. always i hope this time it is working and then after some days again all lost. like he write: all mbr complete gone, how can the software have so many rights to destroy a full partition. 

_sticks are_ 
sandisk cruizer fit
Transcend JetFlash 
and Sandisk ultra

**Reply by keirf** [reply]
File selector sometimes writes the mbr (block zero) on some systems. It seems a feature of particular Amigas, issue has not been tracked down.

Use native selection if you can (not convenient without a better display than standard three digit unfortunately)

**Comment by keirf** [comment]
All said and done there's probably no good reason to allow block 0 writes. I can't think of one. Are you interested in testing a firmware which blocks such writes, or perhaps block writes until an explicit LBA change after entering direct access mode?

**Reply by manga303** [reply]
yes i would like to have such firmware.

**Reply by nicodex** [reply]
> All said and done there's probably no good reason to allow block 0 writes. I can't think of one.

Personally I would like to see this 'security feature' reverted - or at least give us a way to disable it, since we know what we are doing :)

Protecting MBR writes might look like a good idea, but limiting writes to the active FF FAT partition prevents/breaks a lot of usage scenarios.

I'm using the D-A mode to R/W access a second partition with a custom Amiga device driver. I do not want to develop another FAT driver that is already in the firmware, and I do not want to care about synchronizing the file system changes with the firmware (prohibits caching on the host side)...

**Reply by keirf** [reply]
I could give you a magic command to unlock this. It just needs to be implausible to unlock accidentally. It could re-lock every time the D-A track is re-entered.

**Reply by nicodex** [reply]
> I could give you a magic command to unlock this. It just needs to be implausible to unlock accidentally. It could re-lock every time the D-A track is re-entered.

I'm not sure if the auto-lock would work that good. The D-A interface already costs a lot of performance. Using it with an active AmigaOS, the only clean way is using the disk.resource/GetUnit interface to access the device in sync with the trackdisk.device, which forces us to restore the track/head position. Unlocking every D-A cycle would add another costly wait on the index and write encoding+DMA.

Therefore I would prefer a config option to disable this feature completely and/or partially (there are cases where I ant to write the MBR, since some of my sticks/cards have a dual MBR+RDB that are kept in sync).
Or both (config and unlock). Whatever you prefer, I will get used to it :-)

**Reply by keirf** [reply]
One thing that may speed you up: Are you aware that you can write sectors with a special header that includes the sector number? Rather than writing at the exact rotational position?

**Reply by keirf** [reply]
I may be amenable to a FF.CFG option I guess.

**Reply by keirf** [reply]
Finally, is this new Amiga device driver available for download? Open source? Is it on GitHub?

**Reply by nicodex** [reply]
> Are you aware that you can write sectors with a special header that includes the sector number? Rather than writing at the exact rotational position?

No, didn't even thought about this yet (the old/original spec explicitly mentions the index signal).
As of now I used the HxC firmware and have to test if that is also working there...

**Reply by nicodex** [reply]
> Finally, is this new Amiga device driver available for download? Open source? Is it on GitHub?

Would require a lot of cleanup and polishing before going public (unlikely to happen this or next month). When done, it will be on github, yes.

**Comment by keirf** [comment]
Okay, here is a test firmware which disallows writes outside of the mounted FAT volume. This will disallow writes to the MBR, and even to the mounted volume's own boot/BPB sector. It should prevent total trashing of the pen drive. You need to be logged into Github to follow this link:
https://github.com/keirf/flashfloppy/suites/6609057357/artifacts/248564501

**Reply by keirf** [reply]
@manga303 
@slajerek

**Reply by manga303** [reply]
thank you very much ! i update to this and it looks good, i will just use it now for a week in everyday use and report back.

**Reply by slajerek** [reply]
I wonder if this has been already fixed/merged? I have an error "Failed to generate URL to download artifact."

**Reply by keirf** [reply]
Yes this is merged and fixed in latest FlashFloppy releases.
